service/worker/domainReplicationTaskHandler.go

```diff
@@ -0,0 +1,204 @@
+	// ErrInvalidDomainInfo is the error to indicate empty info attribute
+	ErrInvalidDomainInfo = errors.New("invalid domain info attribute")
+	// ErrInvalidDomainConfig is the error to indicate empty config attribute
+	ErrInvalidDomainConfig = errors.New("invalid domain config attribute")
+	// ErrInvalidDomainReplicationConfig is the error to indicate empty replication config attribute
+	ErrInvalidDomainReplicationConfig = errors.New("invalid domain replication config attribute")
+	// ErrInvalidDomainFailoverVersion is the error to indicate empty failover version attribute
+	ErrInvalidDomainFailoverVersion = errors.New("invalid domain failover version attribute")
+	// ErrInvalidDomainStatus is the error to indicate invalid domain status
+	ErrInvalidDomainStatus = errors.New("invalid domain status attribute")
+)
+
+type (
+	domainReplicatorImpl struct {
+		metadataManager persistence.MetadataManager
+		logger          bark.Logger
+	}
+)
+
+// newDomainReplicator create a new instance odf domain replicator
+func newDomainReplicator(metadataManager persistence.MetadataManager, logger bark.Logger) DomainReplicator {
+	return &domainReplicatorImpl{
+		metadataManager: metadataManager,
+		logger:          logger,
+	}
+}
+
+// handleDomainReplicationTask handle the domain replication task
+func (domainReplicator *domainReplicatorImpl) HandleTask(task *replicator.DomainTaskAttributes) error {
+	if err := domainReplicator.validateDomainReplicationTask(task); err != nil {
+		return err
+	}
+
+	switch task.GetDomainOperation() {
+	case replicator.DomainOperationCreate:
+		return domainReplicator.handleDomainCreationReplicationTask(task)
+	case replicator.DomainOperationUpdate:
+		return domainReplicator.handleDomainUpdateReplicationTask(task)
+	default:
+		return ErrInvalidDomainOperation
+	}
+}
+
+// handleDomainCreationReplicationTask handle the domain creation replication task
+func (domainReplicator *domainReplicatorImpl) handleDomainCreationReplicationTask(task *replicator.DomainTaskAttributes) error {
+	// task already validated
+
+	status, err := domainReplicator.convertDomainStatus(task.Info.Status)
+	if err != nil {
+		return err
+	}
+
+	request := &persistence.CreateDomainRequest{
+		Info: &persistence.DomainInfo{
+			ID:          task.GetID(),
+			Name:        task.Info.GetName(),
+			Status:      status,
+			Description: task.Info.GetDescription(),
+			OwnerEmail:  task.Info.GetOwnerEmail(),
+		},
+		Config: &persistence.DomainConfig{
+			Retention:  task.Config.GetWorkflowExecutionRetentionPeriodInDays(),
+			EmitMetric: task.Config.GetEmitMetric(),
+		},
+		ReplicationConfig: &persistence.DomainReplicationConfig{
+			ActiveClusterName: task.ReplicationConfig.GetActiveClusterName(),
+			FailoverVersion:   task.GetFailoverVersion(),
+			Clusters:          domainReplicator.convertClusterReplicationConfig(task.ReplicationConfig.Clusters),
+		},
+	}
+
+	_, err = domainReplicator.metadataManager.CreateDomain(request)
```

Comment:
Can we go through the frontend instead of directly calling the metadata CRUD?

Reply:
on the frontend, there are logics which deal with domain failover, and a lot of input validation.
on the replicator side, the replicator only does basic logic, i.e. save the domain object to DB, with correct DB record version, and comparison of failover version.